lib/linked_list.rb

```diff
@@ -18,73 +18,156 @@ def initialize
     def find_min
-      raise NotImplementedError
+      return nil if @head == nil
+
+      visited_node = @head
+      min = visited_node.data
+
+      while visited_node != nil
+        if visited_node.data  < min
+          min = visited_node.data
+        end
+        visited_node = visited_node.next
+      end
+      return min
     end
 
 
     # method that returns the length of the singly linked list
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity: O(n) - must count all nodes
+    # Space Complexity: O - constant value returned
     def length
-      raise NotImplementedError
+      count = 0
+      current = @head
+
+      if @head == nil
+        return 0
+      else
+        until current == nil
+          current = current.next
+          count += 1
+        end
+      end
+      return count
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity:  O(n) - worst case
+    # Space Complexity: O - constant value returned
     def get_at_index(index)
-      raise NotImplementedError
+      current = @head
+
+      return nil if @head == nil
+      count = 0
+
+      while count < index && current.next != nil
+        current = current.next
+        count += 1
+      end
+
+      return current.data
     end
 
     # method to print all the values in the linked list
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity:  O(n)
+    # Space Complexity O - just printing to console, no new memory needed
     def visit
-      raise NotImplementedError
+      current = @head
+      
+      while current.next != nil
+        puts @current.data
+        current = current.next
+      end
+      # for last node
+      puts current.data
     end
 
     # method to delete the first node found with specified value
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity: O(n)
+    # Space Complexity O
     def delete(value)
-      raise NotImplementedError
+      current = @head
+      
+      if @head == nil
+        return 0
+      end
+
+      # if head node has the value to delete
+      if current.data == value
+        @head = current.next
+      end
+
+      while current!= nil
+        if current.next != nil && current.next.data == value
+          current.next = current.next.next
+        end
+        current = current.next
+      end
     end
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity:  O(n)
+    # Space Complexity O(n)
     def reverse
-      raise NotImplementedError
+      current = @head
+      reversed_list = LinkedList.new
+
+      while current.next != nil
+        reversed_list.add_last(current)

```

Comment:
This isn't working because it is adding the **node** to the new list, rather than the node's data.  It's also adding them all to the last, rather than the front.

Lastly this loop is skipping the last node in the list (check the while condition).